docs/fileformat-prose.md

```diff
@@ -0,0 +1,318 @@
+# On-disk format
+
+```{note}
+These docs are written for anndata 0.8.
+Files written before this version may differ in some conventions,
+but will still be read by newer versions of the library.
+```
+
+AnnData objects are saved on disk to hierarchical array stores like [HDF5]
+(via {doc}`H5py <h5py:index>`) and {doc}`zarr:index`.
+This allows us to have very similar structures in disk and on memory.
+
+As an example we’ll look into a typical `.h5ad` object that’s been through an analysis.
+This structure should be largely equivalent to Zarr structure, though there are a few minor differences.
+
+## Elements
+
+
+ <!-- I’ve started using h5py since I couldn’t figure out a nice way to print attributes from bash. -->
+
+```python
+>>> import h5py
+>>> f = h5py.File("02_processed.h5ad", "r")
+>>> list(f.keys())
+['X', 'layers', 'obs', 'obsm', 'uns', 'var', 'varm']
+```
+
+<!-- ```bash
+$ h5ls 02_processed.h5ad
+X                        Group
+layers                   Group
+obs                      Group
+obsm                     Group
+uns                      Group
+var                      Group
+varm                     Group
+``` -->
+
+In general, `AnnData` objects are comprised of a various types of elements.
+Each element is encoded as either an Array (or Dataset in hdf5 terminology) or a collection of elements (e.g. Group) in the store.
+We record the type of an element using the `encoding-type` and `encoding-version` keys in it's attributes.
+For example, we can this file represents an `AnnData` object from this metadata:
+
+```python
+>>> dict(f.attrs)
+{'encoding-type': 'anndata', 'encoding-version': '0.1.0'}
+```
+
+Using this information, we're able to dispatch onto readers for the different element types that you'd find in an anndata.
+
+### Element Specification
+
+* An element can be any object within the storage hierarchy (typically an array or group) with associated metadata
+* An element MUST have a string valued field `"encoding-type"` in it's metadata
+* An element MUST have a string valued field `"encoding-version"` in it's metadata that can be evaluated to a version
+
+### AnnData specification (v0.1.0)
+
+* An `AnnData` object MUST be a group. 
+* The group's metadata MUST include entries: `"encoding-type": "anndata"`, `"encoding-version": "0.1.0"`.
+* An `AnnData` group MUST contain entries `"obs"` and `"var"`, which MUST be dataframes.
```

Comment:
yet they can be empty dataframes but with specified index? I would specify that

Reply:
```suggestion
* An `AnnData` group MUST contain entries `"obs"` and `"var"`, which MUST be dataframes (though this may only have an index with no columns).
```